File: sigtrack/src/rangelist.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include "rangelist.h"

///////////////////////////////////////////////////////////////////////////////
//
// Copyright 2015-2022, Population Health Research Institute
// Copyright 2015-2022, Martin Renters
//
// This file is part of the DataFax Toolkit.
//
// The DataFax Toolkit is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// The DataFax Toolkit is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with The DataFax Toolkit.  If not, see <http://www.gnu.org/licenses/>.
//
///////////////////////////////////////////////////////////////////////////////

/////////////////////////////////////////////////////////////////////////
// Rangelist List Support Functions
//
// Author: Martin Renters, Oct, 2015
/////////////////////////////////////////////////////////////////////////

/////////////////////////////////////////////////////////////////////////
// rl_addtofront - Adds a new range to the front of a RangeList
/////////////////////////////////////////////////////////////////////////
RangeList *rl_addtofront(RangeList *l, long long min, long long max)
{
	RangeList *l0;

	l0 = malloc(sizeof(RangeList));
	if (min > max) {	// Swap min/max
		min ^= max;
		max ^= min;
		min ^= max;
	}
	l0->next = l;
	l0->min = min;
	l0->max = max;
	return l0;
}

/////////////////////////////////////////////////////////////////////////
// rl_dup - Duplicate a RangeList
/////////////////////////////////////////////////////////////////////////
RangeList *rl_dup(RangeList *l)
{
	RangeList *n, *n0;
	
	n = n0 = 0;

	for (; l; l=l->next) {
		if (!n) {
			n0 = malloc(sizeof(RangeList));
			n = n0;
		} else {
			n0->next = malloc(sizeof(RangeList));
			n0 = n0->next;
		}
		n0->next = 0;
		n0->min = l->min;
		n0->max = l->max;
	}
	return n;
}

/////////////////////////////////////////////////////////////////////////
// rl_free - Free a RangeList
/////////////////////////////////////////////////////////////////////////
void rl_free(RangeList *l)
{
	RangeList *l0;

	while (l) {
		l0 = l->next;
		free(l);
		l = l0;
	}
}

/////////////////////////////////////////////////////////////////////////
// rl_min - Find minimum value in rangelist
/////////////////////////////////////////////////////////////////////////
long long rl_min(RangeList *l)
{
	long long v = 0, first = 1;
	for (;l; l=l->next) {
		if (first || (v > l->min)) {
			v = l->min;
			first = 0;
		}
	}
	return v;
}

/////////////////////////////////////////////////////////////////////////
// rl_max - Find maximum value in rangelist
/////////////////////////////////////////////////////////////////////////
long long rl_max(RangeList *l)
{
	long long v = 0, first = 1;
	for (;l; l=l->next) {
		if (first || (v < l->max)) {
			v = l->max;
			first = 0;
		}
	}
	return v;
}

/////////////////////////////////////////////////////////////////////////
// rl_width - Find number of values covered in rangelist
/////////////////////////////////////////////////////////////////////////
long long rl_width(RangeList *l)
{
	long long v = 0;
	for (;l; l=l->next) {
		v = v + (l->max - l->min) + 1;
	}
	return v;
}

/////////////////////////////////////////////////////////////////////////
// rl_contains - Check whether a RangeList contains a value
/////////////////////////////////////////////////////////////////////////
bool rl_contains(RangeList *l, long long v)
{
	for (;l;l=l->next) {
		if ((v >= l->min) && (v <= l->max)) return true;
	}
	return false;
}

/////////////////////////////////////////////////////////////////////////
// rl_tostring - Convert a RangeList to a string
/////////////////////////////////////////////////////////////////////////
char *rl_tostring(RangeList *l)
{
	int n,len = 1;
	char *s;
	RangeList *l0;

	for (l0=l; l0; l0=l0->next) {
		len += snprintf(NULL, 0, "%lld", l0->min);
		if (l0->min != l0->max) {		// Dash + max
			len += snprintf(NULL, 0, "-%lld", l0->max);
		}
		if (l0->next) len+=1;		// For comma
	}

	s = malloc(len);
	n = 0;
	for (l0=l; l0; l0=l0->next) {
		n += snprintf(s+n, len-n, "%lld", l0->min);
		if (l0->min != l0->max) {		// Dash + max
			n += snprintf(s+n, len-n, "-%lld", l0->max);
		}
		if (l0->next) *(s+(n++)) = ','; 
	}
	*(s+n) = 0;

	return s;
}

typedef enum { NUMBER, COMMA, DASH, ERROR, EOL } Token;

/////////////////////////////////////////////////////////////////////////
// rl_fromstring - Convert a string to RangeList
/////////////////////////////////////////////////////////////////////////
int rl_fromstring(RangeList **rl, char *s)
{
	long long v;
	RangeList *l=0;
	Token token, last;

	if (!s || (*s==0)) {		// Empty string
		*rl = 0;
		return 0;		// No error
	}

	if (strcmp(s, "*")==0) {	// Wildcard
		l = malloc(sizeof(RangeList));
		l->next = 0;
		l->min = 0;
		l->max = 0x7FFFFFFF;
		*rl = l;
		return 0;		// No Error
	}

	last = NUMBER;
	for (;*s;s++) {
		if (isspace(*s)) continue;
		if (isdigit(*s)) {
			v = 0;
			for (;isdigit(*s);s++) {
				v = (v*10) + (*s-'0');
			}
			token=NUMBER;
			s--;		// Back up one
		} else if (*s==',') {
			token=COMMA;
		} else if (*s=='-') {
			token=DASH;
		} else if (*s==0) {
			token=EOL;
		} else {
			token=ERROR;
		}

		switch(token) {
		case NUMBER:
			if (last == DASH) {
				if (l) {
					l->max = v;
					if (l->min > l->max) {
						l->min ^= l->max;
						l->max ^= l->min;
						l->min ^= l->max;
					}
					l = 0;
				} else {
					return -1;	// Error, DASH no digit
				}
			} else {
				l = malloc(sizeof(RangeList));
				l->next = 0;
				l->min = l->max = v;
				*rl = l;
				rl = &l->next;
			}
			break;
		case ERROR:
			return -1;
		case COMMA:
			l = 0;
			break;
		case DASH:
		case EOL:
			break;
		}
		last = token;
	}

	if (last != NUMBER) return -1;

	return 0;
}

#ifdef DEBUG_MAIN
int main(int argc, char *argv[])
{
	char *s;
	RangeList *l = 0;

	l=rl_addtofront(l, 1, 10);
	l=rl_addtofront(l, 5, 5);
	l=rl_addtofront(l, 100, 200);
	s=rl_tostring(l);
	printf("%d=%s\n", strlen(s), s);
	printf("fromstring=%d\n", rl_fromstring(&l, argv[1]));
	s=rl_tostring(l);
	printf("%d=%s\n", strlen(s), s);
}
#endif

File: dftoolkit/rangelist.py
# vim:ts=4:sw=4:et:ai:sts=4
#
# Copyright 2020-2025, Martin Renters
#
# This file is part of DFtoolkit
#
# DFtoolkit is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# DFtoolkit is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with DFtoolkit.  If not, see <http://www.gnu.org/licenses/>.
#

'''RangeList functions and support classes'''

import argparse
import re

class RangeList:
    '''
    RangeList provides a class for storing and manipulating range lists in the
    form of: number, number-number
    '''
    def __init__(self, min_value, max_value, default_all=False):
        '''Initialize a Rangelist'''
        self.min_value = min_value
        self.max_value = max_value
        self.default_all = default_all
        self.values = []

    @property
    def empty(self):
        '''returns whether the rangelist has entries'''
        return len(self.values) == 0

    def from_string(self, range_string):
        '''Convert a range list string to a rangelist item'''
        self.values = []

        # Check for ALL keyword
        if range_string.strip().upper() == 'ALL':
            self.append(self.min_value, self.max_value)
            return

        # Remove beginning and trailing spaces
        range_string = range_string.strip()
        range_string = range_string.replace('~', '-')
        range_string = re.sub(r'[ ]+-', '-', range_string)
        range_string = re.sub(r'-[ ]+', '-', range_string)
        range_string = re.sub(r'[ ]+', ',', range_string)
        range_string = re.sub(r',+', ',', range_string)
        range_items = range_string.split(',')
        for range_item in range_items:
            if range_item == '':
                continue
            if range_item == '*':
                self.append(self.min_value, self.max_value)
                continue

            range_item_list = range_item.split('-')
            if len(range_item_list) == 2:
                self.append(int(range_item_list[0]),
                            int(range_item_list[1]))
            elif len(range_item_list) == 1:
                self.append(int(range_item_list[0]),
                            int(range_item_list[0]))
            else:
                raise ValueError('Invalid range specification')

    def append(self, low, high):
        '''Appends a low, high value to a RangeList'''
        if low > high or low < self.min_value or low > self.max_value or \
                high < self.min_value or high > self.max_value:
            raise ValueError('Invalid range specification')

        self.values.append((low, high))

    def __contains__(self, value):
        '''Returns whether value appears in the list'''
        if not self.values:
            return self.default_all

        for range_value in self.values:
            if range_value[0] <= value <= range_value[1]:
                return True
        return False

    def position(self, value):
        '''Returns the position in the list where value is located'''
        pos = -1
        for pos, range_value in enumerate(self.values):
            if range_value[0] <= value <= range_value[1]:
                return pos
        return pos+1

    def to_string(self):
        '''Returns a string representation of a RangeList'''
        return ','.join(
            ['{0}'.format(r[0]) if r[0] == r[1] else \
             '{0}-{1}'.format(r[0], r[1]) for r in self.values])

    def __repr__(self):
        return self.to_string()

    def __str__(self):
        return self.to_string()

    def sql(self, field_name):
        '''Generate an SQL clause from the rangelist'''
        if not self.values:
            return None
        clause = ''
        for range_value in self.values:
            if clause:
                clause += ' or '
            if range_value[0] == range_value[1]:
                clause += field_name + '=' + str(range_value[0])
            else:
                clause += field_name + ' between ' + str(range_value[0]) + \
                            ' and ' + str(range_value[1])
        return '(' + clause + ')'


class SiteList(RangeList):
    '''A representation of a list of sites'''
    def __init__(self, **kwargs):
        super(SiteList, self).__init__(0, 21460, **kwargs)

class SubjectList(RangeList):
    '''A representation of a list of patients'''
    def __init__(self, **kwargs):
        super(SubjectList, self).__init__(1, 281474976710656, **kwargs)

class PlateList(RangeList):
    '''A representation of a list of plates'''
    def __init__(self, **kwargs):
        super(PlateList, self).__init__(1, 500, **kwargs)

class VisitList(RangeList):
    '''A representation of a list of visits'''
    def __init__(self, **kwargs):
        super(VisitList, self).__init__(0, 65535, **kwargs)

class LevelList(RangeList):
    '''A representation of a list of levels'''
    def __init__(self, **kwargs):
        super(LevelList, self).__init__(0, 7, **kwargs)

class RangeListAction(argparse.Action):
    '''
    An argparse action to create a RangeList. Example:
    parser.add_argument('--ids', default=SubjectList(default_all=True),
                        action=RangeListAction)
    '''
    def __init__(self, option_strings, dest, nargs=None, **kwargs):
        super(RangeListAction, self).__init__(option_strings, dest, **kwargs)
        if not isinstance(self.default, RangeList):
            raise ValueError('No default rangelist specified')

    def __call__(self, parser, namespace, values, option_string=None):
        range_list = getattr(namespace, self.dest)
        try:
            range_list.from_string(values)
        except:
            raise argparse.ArgumentError(self, 'invalid range: ' + values)
